File: docs/build-log/history-1d6/round5/mutations.mjs
import assert from "node:assert/strict";
import { readFileSync, writeFileSync } from "node:fs";
import { resolve } from "node:path";
import { spawnSync } from "node:child_process";

const root = resolve(import.meta.dirname, "../../../..");
const endpoint = "mcp/src/clinical-graph/hpi-endpoint.ts";
const hook = "ui/src/components/charting/useHistoryItemReview.tsx";
function replaceOnce(source, before, after) {
  assert.equal(source.split(before).length, 2, `Mutation anchor must occur once: ${before}`);
  return source.replace(before, after);
}
const cases = [
  { id: 1, name: "oversize-answer-unit", file: endpoint, cwd: "mcp", test: "tests/historyRosHttp.test.ts", pattern: "bulk units derive seven",
    mutate: source => replaceOnce(source, "const answersPerUnit = HISTORY_CONDITIONAL_BUNDLE_ENTRY_LIMIT - 1;", "const answersPerUnit = HISTORY_CONDITIONAL_BUNDLE_ENTRY_LIMIT;") },
  { id: 3, name: "intended-act-targets", file: endpoint, cwd: "mcp", test: "tests/historyRosHttp.test.ts", pattern: "conditional create preserves",
    mutate: source => replaceOnce(source, "recordHistoryItemReview(staff.fhir, reviewInput(liveNegativeTargets))", "recordHistoryItemReview(staff.fhir, reviewInput(request.targets))") },
  { id: 4, name: "remove-answer-conditional-create", file: endpoint, cwd: "mcp", test: "tests/historyRosHttp.test.ts", pattern: "conditional create preserves",
    mutate: source => replaceOnce(source, "ifNoneExist: `identifier=${HISTORY_ANSWER_IDENTIFIER_SYSTEM}|${answer.id}&status:not=entered-in-error&status:not=cancelled`,", "ifNoneExist: undefined,") },
  { id: 5, name: "remove-completed-act-short-circuit", file: endpoint, cwd: "mcp", test: "tests/historyRosHttp.test.ts", pattern: "completed gesture replay",
    mutate: source => replaceOnce(source, "if (await findHistoryItemAct(staff.fhir, actIdentifier)) {", "if (false && await findHistoryItemAct(staff.fhir, actIdentifier)) {") },
  { id: 6, name: "restore-retired-summary-token", file: "mcp/src/clinical-graph/history-template-engine.ts", cwd: "mcp", test: "tests/historyRos.test.ts", pattern: "prior encounter acts",
    mutate: source => replaceOnce(source, 'summary: "{reviewed_systems}. {positives}",', 'summary: "{reviewed_systems}. {positives}. {method}",') },
  { id: 11, name: "remove-legacy-bulk-refusal", file: endpoint, cwd: "mcp", test: "tests/historyRosHttp.test.ts", pattern: "I2 every",
    mutate: source => replaceOnce(source, "return handleHistoryItemReviewRequest(deps, input);", "return historySearchResult(() => handleHistoryReview(deps, input));") },
  { id: 14, name: "remove-row-freeze", file: "ui/src/components/charting/HistoryRosSection.tsx", cwd: "ui", test: "tests/historyRosBrowser.test.tsx", pattern: "comprehensive burst",
    mutate: source => replaceOnce(source, "<fieldset disabled={sectionBusy}", "<fieldset disabled={false}") },
  { id: 15, name: "divergent-http-door", file: endpoint, cwd: "mcp", test: "tests/historyRosHttp.test.ts", pattern: "I2 every",
    mutate: source => replaceOnce(source, "return handleHistoryItemReviewRequest(deps, input);", "return historySearchResult(() => handleHistoryReview(deps, input));") },
  { id: 18, name: "restore-stale-target-fallback", file: hook, cwd: "ui", test: "tests/historyRosBrowser.test.tsx", pattern: "comprehensive burst",
    mutate: source => {
      source = replaceOnce(source, "  const owner = useRef<object>();", "  const owner = useRef<object>();\n  const bulkProgress = useRef<Target[]>();");
      return replaceOnce(source, "    if (!targets.length) return undefined;", "    targets = bulkProgress.current ?? targets;\n    bulkProgress.current = targets;\n    if (!targets.length) return undefined;");
    } },
  { id: 19, name: "unconditional-lease-release", file: hook, cwd: "ui", test: "tests/historyRosBrowser.test.tsx", pattern: "comprehensive burst",
    mutate: source => replaceOnce(source, "if (ownsRequest()) { release(); pending.current = false; setBusy(false); }", "release();\n      if (ownsRequest()) { pending.current = false; setBusy(false); }") },
  { id: 20, name: "client-derived-act-targets", file: endpoint, cwd: "mcp", test: "tests/historyRosHttp.test.ts", pattern: "conditional create preserves",
    mutate: source => replaceOnce(source, "recordHistoryItemReview(staff.fhir, reviewInput(liveNegativeTargets))", "recordHistoryItemReview(staff.fhir, reviewInput(request.targets))") },
];

function run(probe, phase) {
  const args = ["--import", "tsx", "--test", `--test-name-pattern=${probe.pattern}`, probe.test];
  const result = spawnSync(process.execPath, args, { cwd: resolve(root, probe.cwd), encoding: "utf8", timeout: 90000, maxBuffer: 8 * 1024 * 1024 });
  const output = result.stdout + result.stderr;
  writeFileSync(resolve(import.meta.dirname, `${probe.id}-${phase}.log`), output);
  const count = name => Number(output.match(new RegExp(`^# ${name} (\\d+)$`, "m"))?.[1] ?? NaN);
  return { exit: result.status, signal: result.signal, passed: count("pass"), failed: count("fail"), skipped: count("skipped") };
}

const results = [];
for (const probe of cases) {
  const path = resolve(root, probe.file);
  const original = readFileSync(path, "utf8");
  let red;
  try {
    writeFileSync(path, probe.mutate(original));
    red = run(probe, "red");
  } finally {
    writeFileSync(path, original);
  }
  const restored = run(probe, "restored");
  results.push({ id: probe.id, name: probe.name, file: probe.file,
    command: `cd ${probe.cwd} && node --import tsx --test --test-name-pattern='${probe.pattern}' ${probe.test}`, red, restored });
  writeFileSync(resolve(import.meta.dirname, "results.json"), JSON.stringify(results, null, 2) + "\n");
  console.log(JSON.stringify(results.at(-1)));
  assert.equal(red.exit, 1); assert.equal(red.failed, 1); assert.equal(red.passed, 0);
  assert.equal(restored.exit, 0); assert.equal(restored.failed, 0); assert.equal(restored.passed, 1);
}

File: docs/build-log/followup-s3b2-picker/proof/mutations.py
from pathlib import Path
import json, subprocess, shlex
root=Path(__file__).resolve().parents[4]
store='mcp/src/clinical-graph/exam-scope-store.ts'
endpoint='mcp/src/clinical-graph/exam-overview-endpoint.ts'
picker='ui/src/components/charting/FollowingPicker.tsx'
server=lambda pattern,file: ['node','--import','tsx','--test','--test-name-pattern='+pattern,file]
ui=lambda pattern,file: ['node','--import','tsx','--test','--test-name-pattern='+pattern,file]
cases=[
 ('G1',store,'if (existing) return parseScope(existing);','',server('S3b2 G1','mcp/tests/examShapeRecord.test.ts'),'.'),
 ('G2',store,'const chosenAt = new Date().toISOString();','if (existing) throw new Error("mutation refuses existing row");\n    const chosenAt = new Date().toISOString();',server('S3b2 G2','mcp/tests/examShapeRecord.test.ts'),'.'),
 ('G3',store,'if (existing) return parseScope(existing);','if (existing) return { ...parseScope(existing), sectionsOpen: freezeProfiles(await resolve()).sectionsOpen };',server('S3b2 G3','mcp/tests/examShapeRecord.test.ts'),'.'),
 ('G4',store,'z.enum(["derived", "explicit"]).default("derived")','z.enum(["derived", "explicit"])',server('S3b2 G4','mcp/tests/examShapeRecord.test.ts'),'.'),
 ('G5',endpoint,'scope = await store.pick(encounterId,','await staff.fhir.create({ resourceType: "Condition", subject: encounter.subject });\n        scope = await store.pick(encounterId,',server('S3b2 G1 G2 G5','mcp/tests/examOverviewEndpoint.test.ts'),'.'),
 ('G6',picker,'useState<FollowingChoice["examScope"]>("office-visit")','useState<FollowingChoice["examScope"]>((globalThis as any).appointment?.category === "comprehensive" ? "comprehensive" : "office-visit")',ui('S3b2 G6','tests/followingPicker.test.tsx'),'ui'),
 ('G7',endpoint,'// Opening the board must not depend on bookkeeping persistence.','throw error;\n        // Opening the board must not depend on bookkeeping persistence.',server('S3b1 G8','mcp/tests/examOverviewEndpoint.test.ts'),'.'),
 ('G8','ui/tests/clinicalGraphRouting.test.tsx','assert.equal(callers.length, 57);','assert.equal(callers.length, 56);',ui('clinical-graph requests share','tests/clinicalGraphRouting.test.tsx'),'ui'),
]
results=[]
def run(command,cwd):
 p=subprocess.run(command,cwd=root/cwd,text=True,capture_output=True)
 summary='\n'.join(line for line in (p.stdout+'\n'+p.stderr).splitlines() if line.startswith(('ok ','not ok ','# tests ','# pass ','# fail ','# skipped ')))
 return {'exit':p.returncode,'summary':summary}
for guard,file,before,after,command,cwd in cases:
 path=root/file; original=path.read_text()
 assert before in original, guard
 try:
  path.write_text(original.replace(before,after,1))
  red=run(command,cwd)
 finally: path.write_text(original)
 green=run(command,cwd)
 row={'guard':guard,'file':file,'command':shlex.join(command),'cwd':cwd,'red':red,'green':green}; results.append(row)
 print(guard, 'red',red['exit'],'green',green['exit'],flush=True)
 (root/'docs/build-log/followup-s3b2-picker/mutations.json').write_text(json.dumps(results,indent=2)+'\n')
 assert red['exit'] != 0 and green['exit'] == 0, row
